src/track/external-captions-handler.js

```diff
@@ -0,0 +1,397 @@
+import Error from '../error/error'
+import * as Utils from '../utils/util'
+import {Parser, StringDecoder} from './text-track-display'
+import {TrackType} from "./track-type";
+import TextTrack from "./text-track";
+import {CustomEventType, Html5EventType} from "../event/event-type";
+import FakeEvent from "../event/fake-event";
+import getLogger from "../utils/logger";
+import EventManager from "../event/event-manager";
+
+class ExternalCaptionsHandler {
+
+  static SRT_POSTFIX: string = 'srt';
+  /**
+   * The player class logger.
+   * @type {any}
+   * @static
+   * @private
+   */
+  static _logger: any = getLogger('ExternalCaptionsHandler');
+  /**
+   * Index that specifies the last cue that is playing / played in the text track cue array.
+   * @type {number}
+   * @private
+   */
+  _externalCueIndex: number = 0;
+  /**
+   * a callback of the timeupdate, for external captions handler.
+   * @type {Function}
+   * @private
+   */
+  _handleExternalCaptionsCallback: Function = null;
+  /**
+   * the player object.
+   * @type {Object}
+   * @private
+   */
+  _player: null;
+  /**
+   * event manager for the external caption handler
+   * @type {EventManager}
+   * @private
+   */
+  _eventManager: EventManager = null;
+
+  /**
+   * a map that holds the current cues that are in process. process may be in download or that the cues are being parsed.
+   * @type {Object} language: promise<*>
+   * @private
+   */
+  _cuesInProcessMap: Object = {};
+
+  /**
+   * constructor
+   * @param {Object} player - the player object.
+   */
+  constructor(player: Object) {
+    this._player = player;
+    this._eventManager = new EventManager();
+    this._addListeners();
+  }
+
+  /**
+   * adds listeners to the player, so the module will handle them on its own.
+   * @returns {void}
+   * @private
+   */
+  _addListeners(): void {
+    this._eventManager.listen(this._player, this._player.Event.SEEKED, (e) => {
+      this._maybeSetExternalCueIndex(e)
+    });
+    this._eventManager.listen(this._player, this._player.Event.TEXT_TRACK_CHANGED, (e) => {
+      this._handleTextTrackChanged(e)
+    });
+  }
+
+  /**
+   * Make a request to download a caption and parse it's content to cues.
+   * @param {TextTrack} textTrack - download and parse the cues of the text track
+   * @returns {Promise<any>} - resolves when the request returns and the caption string is parsed to cues.
+   * @private
+   */
+  _getCuesArray(textTrack: TextTrack): Promise<*> {
```

Comment:
I think this method does more then it should, in general the name doesn't imply what it does.
This method downloads a file, tries to parse it and then create and return a text cue list.
I would prefer you break it down to:
```
this._downloadTextFile().then(this._parseCues).then(this._updateTextTrackCues)